### PR TITLE
impl(bigtable): add RetryContext aka the cookie jar

### PR DESCRIPTION
**Pull request by dbolduc**
Part of the work for #13447

- - -
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/13448)

### CODE REVIEW

**Review by alevenberg — DISMISSED** [review]

File: google/cloud/bigtable/internal/retry_context.cc [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+
+#include "google/cloud/bigtable/internal/retry_context.h"
+#include "absl/strings/match.h"
+
+namespace google {
+namespace cloud {
+namespace bigtable_internal {
+GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_BEGIN
+
+void RetryContext::PreCall(grpc::ClientContext& context) const {
```
[the remark this answers is not in the document]
Comment (dbolduc) [reply google/cloud/bigtable/internal/retry_context.cc:23]: Yeah... all this class does is inject and extract metadata. So your names make sense.

I was thinking about this as a precursor to a more generic interface which might also do things like `context.set_deadline()`. :shrug: 

Like a:

```cc
namespace google::cloud {

class ConfigureContext {
 public:
  virtual void PreCall(grpc::ClientContext& context) const = 0;
  virtual void PostCall(grpc::ClientContext const& context) = 0;
};

}

struct ConfigureContextOption {
  using Type = std::shared_ptr<ConfigureContext>;
};

} // namespace
```

Gonna leave it as is, mostly out of laziness.

File: google/cloud/bigtable/internal/retry_context.h [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+ *       auto result = stub.Foo(context, request);
+ *       retry_context.PostCall(context);
+ *       return result;
+ *     }, ...);
+ * }
+ * @endcode
+ */
+class RetryContext {
+ public:
+  void PreCall(grpc::ClientContext& context) const;
```
[the remark this answers is not in the document]
Comment (dbolduc) [reply google/cloud/bigtable/internal/retry_context.h:55]: Done.

**Review by scotthart — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/13448)* status: 0 of 6 files reviewed, 4 unresolved discussions (waiting on @alevenberg and @dbolduc)

___
*[`google/cloud/bigtable/internal/retry_context.cc` line 38 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/13448#-NoIPQa72zDP_9vZeDDI:-NoIPQa72zDP_9vZeDDJ:bl7725e) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/6e6f79a5a713312e01c27804c4fd9fdfbb37a0e2/google/cloud/bigtable/internal/retry_context.cc#L38)):*
> ```cpp
>   for (auto const& kv : metadata) {
>     auto key = std::string{kv.first.data(), kv.first.size()};
>     if (absl::StartsWith(key, "x-goog-cbt-cookie")) {
> ```

Just curious, should this check for `x-goog-cbt-cookie` or `x-goog-cbt-cookie-` ?

File: google/cloud/bigtable/internal/retry_context.cc (hunk 1 above)
[the remark this answers is not in the document]
Comment (alevenberg) [reply google/cloud/bigtable/internal/retry_context.cc:23]: If it's gonna be more generic sounds fine.

File: google/cloud/bigtable/internal/retry_context_test.cc [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+    retry_context.PreCall(c2);
+    return metadata_fixture_.GetMetadata(c2);
+  }
+};
+
+TEST_F(RetryContextTest, StartsEmpty) {
+  RetryContext retry_context;
+
+  grpc::ClientContext c;
+  retry_context.PreCall(c);
```
[the remark this answers is not in the document]
Comment (alevenberg) [reply google/cloud/bigtable/internal/retry_context_test.cc:63]: SG

**Review by alevenberg — APPROVED** [review]

**Pull request merged**